lib/compiler/src/v3_core.erl

```diff
@@ -2147,7 +2170,155 @@ annotate_cons(A, H, T, #core{dialyzer=Dialyzer}) ->
             ann_c_cons(A, H, T)
     end.
 
-ubody(B, St) -> uexpr(B, [], St).
+%%%
+%%% Here follows an abstract data structure to help us handle Erlang's
+%%% implicit matching that occurs when a variable is bound more than
+%%% once:
+%%%
+%%%     X = Expr1(),
+%%%     X = Expr2()
+%%%
+%%% What is implicit in Erlang, must be explicit in Core Erlang; that
+%%% is, repeated variables must be eliminated and explicit matching
+%%% must be added. For simplicity, examples that follow will be given
+%%% in Erlang and not in Core Erlang. Here is how the example can be
+%%% rewritten in Erlang to eliminate the repeated variable:
+%%%
+%%%     X = Expr1(),
+%%%     X1 = Expr2(),
+%%%     if
+%%%         X1 =:= X -> X;
+%%%         true -> error({badmatch,X1})
+%%%     end
+%%%
+%%% To implement the renaming, keeping a set of the variables that
+%%% have been bound so far is **almost** sufficient. When a variable
+%%% in the set is bound a again, it will be renamed and a `case` with
+%%% guard test will be added.
+%%%
+%%% Here is another example:
+%%%
+%%%     (X=A) + (X=B)
+%%%
+%%% Note that the operands for a binary operands is allowed to be
+%%% evaluated in any order. Therefore, variables bound on the left
+%%% hand side must not referenced on the right hand side, and vice
+%%% versa. If a variable is bound on both sides, it must be bound
+%%% to the same variable.
+%%%
+%%% Using the simple scheme of keeping track of known variables,
+%%% the example can be rewritten like this:
+%%%
+%%%     X = A,
+%%%     X1 = B,
+%%%     if
+%%%         X1 =:= X -> ok;
+%%%         true -> error({badmatch,X1})
+%%%     end,
+%%%     X + X1
+%%%
+%%% However, this simple scheme of keeping all previously bound variables in
+%%% a set breaks down for this example:
+%%%
+%%%     (X=A) + fun() -> X = B end()
+%%%
+%%% The rewritten code would be:
+%%%
+%%%     X = A,
+%%%     Tmp = fun() ->
+%%%               X1 = B,
+%%%               if
```

Comment:
> I find it rather difficult to understand why the check for the values being the same should be inside the fun. With an initial more naive rewrite, one would first have evaluation of the parts of the expression and then reconcile any bindings that have taken place in both/several branches.

I did think of that, but thought that it would be more complicated. As it turned, my original simple solution turned out to not be that simple in the end.

I think I will have a go at re-implementing my solution. It might turn out to be simpler than the current solution (or miscellaneous complications might make it impracticable or more complicated).

Reply:
@lisztspace @jhogberg I have now made attempt to implement this bug fix in a cleaner way. I will try explain my thought process. This will be a long comment with examples in Core Erlang, so for anyone not comfortable with that:

**TL;DR**: The attempt didn't work out. I will soon merge this pull request in its current form.

Consider this example:
```
bar1(A, B) ->
    Z = [Y = A, Y = B],
    {Y,Z}.
```

The list elements are evaluated in parallel in any order. If the two bindings of `Y` are different, a `badmatch` exception is raised.

The compiler implements that by renaming `Y` the second time it is seen and adding a test that ensuries that the original value and new values are equal:

```
let <Y> = A
in
    case B of
      <Var1> when call 'erlang':'=:='(Var1, Y) ->
          let <Z> =
              [A, B]
          in
              {Y,Z}
      <Var2> when 'true' ->
            primop 'match_fail'({'badmatch',Var2})
    end
```

This pull request does not change this basic behavior, but ensures that the renaming and comparison does not happen in funs when a variable should not be seen.

Thinking of a different approach, I come with this Core Erlang code that actually works:

```
let <Y1, Y2> =
    <let Y = A in Y, let Y = B in Y>
in
    case <> of
      <> when call 'erlang':'=:='(Y1, Y2) ->
          let <Z> =
              [A, B]
          in
              {Y,Z}
      <> when 'true' ->
            primop 'match_fail'({'badmatch',Var2})
    end
```

Here I used the multiple values construct (`#c_values{}`) in Core Erlang. The expression inside `<` and `>` are evaluated in parallel, just as we want.

However, we get into trouble when we'll try do something slightly more complicated, such as:

```
bar2(A, B) ->
    Z = [begin Y = A, a end, begin Y = B, b end],
    {Y,Z}.
```

Since the values construct is a pure syntactic construction, it cannot be nested, so we must flatten the list of values that we will need to bind:

```
let <Y1, E1, Y2, E2> =
    <let Y = A in Y, a, let Y = B in Y, b>
in
    case <> of
      <> when call 'erlang':'=:='(Y1, Y2) ->
          let <Z> =
              [E1, E2]
          in
              {Y,Z}
      <> when 'true' ->
            primop 'match_fail'({'badmatch',Var2})
    end
```

It gets worse:

```
bar3(A, B) ->
    Z = [begin Y = A, Y + 1 end, begin Y = B, Y + 3 end],
    {Y,Z}.
```

When we try to flatten values:

```
<Y = A, Y + 1, Y = B, Y + 3>
```

it won't work, because the bindings are worked out in parallel; the binding of `Y` is not seen when `Y + 1` is evaluated.

This still simple example could be saved with some clever rewriting, but it is difficult to see how we can make it work in the general case.

So it seems that we must rename variables to get this to work in the general case. As currently implemented, the `v3_core` only renames variables when necessary. Alternatively, `v3_core` could rename all variables, keeping a mapping between the old and new names, and a list of (original) variables that are currently in scope. That would be a more massive rewrite of the `v3_core` module and the compiler would have to do more work.